docs/client-specifications/configuration.md

```diff
@@ -0,0 +1,200 @@
+allows future support for things like client-side metrics or other diagnostics helpers.
+
+```typescript
+interface Middleware {
+  wrapRequest(middlewareFn: (GrpcRequest) => GrpcResult): GrpcResult;
+}
+```
+
+### TransportStrategy
+
+The `TransportStrategy` is responsible for configuring network tunables.  We intend to abstract gRPC away from
+users as much as possible, but we will need to provide different gRPC configurations for dev vs. prod environments.
+And ultimately we will need to support advanced customers who need to mess with these knobs.
+
+```typescript
+interface TransportStrategy {
+  maxConcurrentRequests: number;
+  grpcConfig: GrpcConfiguration;
+}
+
+interface GrpcConfiguration {
+  numChannels(): number;
+
+  // These interface functions expose individual settings for gRPC channels.  They are just here to ensure that
+  // strategy implementations provide values for settings that we know to be important.  These may vary by language
+  // since the gRPC implementations in each language have subtly different behaviors.
+  maxSessionMemory(): number;
+  useLocalSubChannelPool(): boolean;
+
+  // This is a map that encapsulates the settings above, and may also include other channel-specific settings.
+  // This allows strategy implementations to provide gRPC config key/value pairs for any available setting, even
+  // if it's not one we've explicitly tried / recommended.  The strategy implementation should implement this by
+  // calling the functions above, along with allowing a mechanism for specifying additional key/value pairs.
+  grpcChannelConfig(): Record<string, string>;
+}
+```
+
+# Advanced Configuration Examples
+
+In this section we expand a bit on how a customer might define their own advanced configuration if they want to
+opt out of our pre-builts.
+
+Here is a sketch of what the implementation of one of the pre-built configurations might look like inside of the
+SDK (users can see the source code for this if they like):
+
+```typescript
+namespace InRegion {
+  export class LowLatency {
+    static latest(): Configuration {
+      return {
+        retryStrategy: new ExponentialBackoffStrategy({
+          maxAttempts: 3,
+          backoffMilliseconds: 0.5,
+          jitterMilliseconds: 0.5,
+          exponentialBackoffFactor: 2,
+        }),
+        middlewares: [], // TODO: this could eventually include some basic client-side metrics OOTB
+        transportStrategy: {
+          maxConcurrentRequests: 2,
+          grpcConfig: new BasicGrpcConfiguration({
+            numChannels: 5
+          })
+        }
+      }
+    }
+  }
+}
+```
+
+If a customer wishes to override the defaults, they can do something like this:
+
+```typescript
+const myConfig = configs.InRegion.LowLatency.latest()
+  .withTransportStrategy({
+    maxConcurrentRequests: 2,
+    grpcConfig: new BasicGrpcConfiguration({
+      numChannels: 5,
+      additionalGrpcChannelSettings: {
+              'foo': 'bar'
```

Comment:
is this only string settings? If can take numbers maybe pass example of overiding that also. If not I think is fine as is.

Reply:
i'm pretty sure it's string->string.